File: src/algorithms/python/Perfect_Squares.py
# -*- coding: utf-8 -*-

from math import sqrt
from Queue import Queue


class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        q = Queue()
        visited = set()
        q.put((n, 0))
        while not q.empty():
            front = q.get()
            upper = int(sqrt(front[0])) + 1
            for i in range(1, upper):
                new_num = front[0] - i * i
                if new_num == 0:
                    return front[1] + 1
                if new_num not in visited:
                    visited.add(new_num)
                    q.put((new_num, front[1] + 1))


if __name__ == "__main__":
    solution = Solution()
    print solution.numSquares(12)
    print solution.numSquares(13)
    print solution.numSquares(9975)

File: src/algorithms/cpp/Perfect_Squares.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <queue>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int numSquares(int n) {
        vector <bool> vis(n + 1, false);
        queue <pair <int, int>> q;
        q.push(make_pair(n, 0));
        while (!q.empty()) {
            pair <int, int> top = q.front();
            q.pop();
            for (int i = int(sqrt(top.first)); i >= 1; -- i) {
                pair <int, int> tmp = make_pair(top.first - i * i, top.second + 1);
                if (tmp.first == 0)     return tmp.second;
                if (!vis[tmp.first]) {
                    vis[tmp.first] = true;
                    q.push(tmp);
                }
            }
        }
        return -1;
    }
};

int main() {
    return 0;
}
